### PR TITLE
ABR 20: Support restore into a different namespace

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**:

==COMMIT_MSG==
The AtlasRestoreService now supports restoring to a different namespace than the one that was backed up.
It is the responsibility of the user to ensure that the data (including Atlas tables) is copied over correctly during the restore.
==COMMIT_MSG==

**Implementation Description (bullets)**:
- API changes in AtlasRestoreService to take RestoreRequests, which include both oldNamespace and newNamespace

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Amended existing tests
- Verified the correct namespaces are used

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**: soon, but we need to unrevert ABR18 before merging this

### CODE REVIEW

File: atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java [hunk 1]
```diff
@@ -163,15 +162,15 @@ public static AtlasRestoreService createForTests(
...
+     * @param restoreRequests the repair requests.
      * @param repairTable supplied function which is expected to repair the given ranges.
      *
      * @return the set of namespaces for which we issued a repair command via the provided Consumer.
      */
     public Set<Namespace> repairInternalTables(
-            Set<Namespace> namespaces, BiConsumer<String, RangesForRepair> repairTable) {
-        Map<Namespace, CompletedBackup> completedBackups = getCompletedBackups(namespaces);
-        Set<Namespace> namespacesToRepair = completedBackups.keySet();
+            Set<RestoreRequest> restoreRequests, BiConsumer<String, RangesForRepair> repairTable) {
```
Comment (gmaretic) [inline atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java:171]: We should probably check somewhere that we are not trying to restore two namespaces into the same one

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStoreTest.java [hunk 2]
```diff
@@ -121,6 +121,6 @@ public void testPartialFailures() throws ExecutionException, InterruptedExceptio
         assertThat(numberOfSuccessfulPue).isBetween(30, 70);
         // too lazy to calculate exactly, rough estimates
         assertThat(numberOfNothingPresent).isBetween(5, 40);
-        assertThat(numberOfValuePresentAfterFailure).isBetween(15, 50);
+        assertThat(numberOfValuePresentAfterFailure).isBetween(10, 55);
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStoreTest.java:124]: more flakes?

File: timelock-api/src/main/conjure/timelock-api.yml [hunk 3]
```diff
@@ -155,7 +155,7 @@ types:
       CompleteRestoreRequest:
         package: com.palantir.atlasdb.backup.api
         fields:
-          completedBackups: set<CompletedBackup>
+          completedBackups: map<Namespace, CompletedBackup>
```
Comment (gmaretic) [inline timelock-api/src/main/conjure/timelock-api.yml:158]: add some docs on what these are?

**Review by gmaretic — APPROVED** [review]
A few nits but looks sensible

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStoreTest.java (hunk 2 above)
Comment (gsheasby) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStoreTest.java:124]: yup -_-

File: atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java (hunk 1 above)
Comment (gsheasby) [reply atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java:171]: Adding a precondition in prepareRestore (and test).

File: timelock-api/src/main/conjure/timelock-api.yml (hunk 3 above)
Comment (gsheasby) [reply timelock-api/src/main/conjure/timelock-api.yml]: done

**Comment by gsheasby** [comment]
Merged as part of #5925.

**Pull request closed without merging**